File: 5-Problem-A-Day/159. Longest Substring with At Most Two Distinct Characters.java
class Solution {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        Map<Character, Integer> ch = new HashMap<>();
        int start = 0, ans = 0;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            ch.put(c, ch.getOrDefault(c, 0) + 1);
            while (ch.size() > 2) {
                int count = ch.get(s.charAt(start));
                if (count == 1)
                    ch.remove(s.charAt(start));
                else
                    ch.put(s.charAt(start), count - 1);
                start++;
            }            
            ans = Math.max(ans, i - start + 1);
        }
        return ans;
    }
}

File: 2021/159. Longest Substring with At Most Two Distinct Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstringTwoDistinct(string s) {
        int disCount = 0, counter[256], start = 0, maxLen = 0;
        fill(counter, counter + 256, 0);
        for (int i = 0; i < s.length(); i++) {
            counter[s[i]]++;
            if (counter[s[i]] == 1) {
                disCount++;
            }
            while (disCount > 2) {
                counter[s[start]]--;
                if (counter[s[start]] == 0) {
                    disCount--;
                }
                start++;
            }
            maxLen = max(maxLen, i - start + 1);
        }
        return maxLen;
    }
};